### ISSUE TITLE
(csharp) Contextual keywords are highlighted incorrectly

### ISSUE THREAD

**Issue by Youssef1313** [highlightjs/highlight.js#3230]
**Describe the issue**

In C#, there are "contextual keywords", which are *not* always treated as keywords. They are highlighted incorrectly in a context where they are not treated as a keyword.

**Sample Code to Reproduce**

```csharp
public class Move
{
    public Move(Point from, Point to)
    {
    }
}
```

**Expected behavior**

`from` isn't highlighted as a keyword.

**Additional context**

In other contexts (LINQ query), `from` should be highlighted as a keyword.

Note: This issue may not be limited to `from`.

**Comment by joshgoebel** [comment]
And in what contexts in from used?

**Comment by Youssef1313** [comment]
@joshgoebel In any expression context. Like `var x = from .....` or `MyMethod(from .....)`.

I have no idea if it's going to be easy to detect if `from` is in an expression context since there is no semantic information available to highlight.js.

Tagging @IEvangelist, in case you can come up with a good way to handle this.

**Comment by joshgoebel** [comment]
Right, that's the problem.  Highlight.js grammars are going to perform poorly with contextual keywords.  We do not understand the full grammar of a language, nor do we desire that grammars in the core library try to do that in excess - it typically results in overly complex and hard to maintain grammars.

Sometimes we can use tricks... you could, for example, only highlight `from` if it follows a `=`... but that's hard without look-behind (which we can't use yet) and when you start needing different tricks for every single contextual keyword then it gets messy very fast.

Perhaps a discussion should be had about whether we should remove contextual keywords?

**Comment by Youssef1313** [comment]
I'm not sure removing contextual keywords is good.
Simple tricks (even if it won't be 100% accurate) couls be better.

For example, if `from` is followed by a comma, we know 100% it is not a keyword.

**Comment by joshgoebel** [comment]
Is there a reason matching `from [ident]` doesn't work?

**Comment by joshgoebel** [comment]
> Simple tricks (even if it won't be 100% accurate) could be better.

We don't need 100%, but it should be close.  Well-written, typical code in a given language should highlight consistently.  So if a trick gets us there it _might_ be ok... but if it only gets us 50% or 80% then we'd prefer not to highlight at all than leave the user guessing when something is or isn't highlighted.

The problem is when you have to hand-code tricks for every single keyword and have lots of contextual keywords...

**Comment by Youssef1313** [comment]
> Is there a reason matching `from [ident]` doesn't work?

Currently this works, and I think it should remain working as long as there is no `,` or `.` or `;` after the indent/space.

---

> The problem is when you have to hand-code tricks for every single keyword and have lots of contextual keywords...

Well, I think the trick I suggested will work with most contextual keywords. `init`, `get`, `set`, `nint`, and `nuint` are the exceptions I can think of, but they are *rarely* used for anything other than being a keyword. So it should be okay to always highlight them as keywords.

---

**Comment by joshgoebel** [comment]
> Currently this works, and I think it should remain working as long as there is no , or . or ; after the indent/space.

No I mean if the rule was `from [indent]` to highlight it... does that catch all value from uses?

**Comment by Youssef1313** [comment]
Putting more thought, both approaches will have a fair false highlighting cases.

For `from` and `join` specifically, it should only be highlighted if it's followed by one or two words then followed by `in`.

For example:

```csharp
from string x in ....
// or
from x in ....
```

However, this violates:

> The problem is when you have to hand-code tricks for every single keyword and have lots of contextual keywords...

But I'm not expecting that there will be lots of tricks. Some contextual keywords are 99.99% used as keywords, not identifiers. So we won't need to bother with large number of tricks for every single contextual keyword.

Side note: there is a language proposal to allow *only* lexical keywords.

**Comment by joshgoebel** [comment]
```
from string x in ....
// or
from x in ....
```

If it's actually always this simple this might be a good use for the new multi-matchers... see `wren.js`, etc...

**Comment by auula** [comment]
I want to develop a code template with automatic typing effects,I used type.js to achieve but lost the highlighting, how can I solve it, I am a rookie programmer😭

**Comment by joshgoebel** [comment]
@Youssef1313 Want to try your hand at a PR?  I feel like we need a special rule just for `from` or `join` that looks ahead and makes sure it has `word word? in` and then remove them from the regular keyword list.

I think `in` could stay in the regular list, correct?

**Comment by joshgoebel** [comment]
Tagging auto-close... we're no longing going to leave hard contextual parsing issues open as these are by their nature impossible to fix fully without a full parser.  If someone would like to work on a PR for the few simple cases, this would be welcome, but it's not a reason for such issues to become zombie issues forever.
